proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,313 @@
+command-line arguments.
+
+The current process involves a lot of human intervention, if people make a
+mistake in deciding the highest optimistically confirmed slot, it is
+detrimental to the viability of the ecosystem.
+
+We aim to automate the negotiation of highest optimistically confirmed slot and
+the distribution of all blocks on that fork, so that we can lower the 
+possibility of human mistakes in the cluster restart process. This also reduces 
+the burden on validator operators, because they don't have to stay around while 
+the validators automatically try to reach consensus, the validator will halt
+and print debug information if anything goes wrong, and operators can set up
+their own monitoring accordingly.
+
+## Alternatives Considered
+
+### Automatically detect outage and perform cluster restart
+
+The reaction time of a human in case of emergency is measured in minutes,
+while a cluster restart where human initiate validator restarts takes hours.
+We considered various approaches to automatically detect outage and perform
+cluster restart, which can reduce recovery speed to minutes or even seconds.
+
+However, automatically restarting the whole cluster seems risky. Because
+if the recovery process itself doesn't work, it might be some time before
+we can get human's attention. And it doesn't solve the cases where new binary
+is needed. So for now we still plan to have human in the loop.
+
+After we gain more experience with the restart approach in this proposal, we
+may slowly try to automate more parts to improve cluster reliability.
+
+### Use gossip and consensus to figure out restart slot before the restart
+
+The main difference between this and the current proposal is this alternative
+tries to make the cluster automatically enter restart preparation phase without 
+human intervention.
+
+While getting human out of the loop improves recovery speed, there are concerns 
+about recovery gossip messages interfering with normal gossip messages, and 
+automatically start a new message in gossip seems risky.
+
+### Automatically reduce block production in an outage
+
+Right now we have vote-only mode, a validator will only pack vote transactions
+into new blocks if the tower distance (last_vote - local_root) is greater than
+400 slots.
+
+Unfortunately in the previous outages vote-only mode isn't enough to save the
+cluster. There are proposals of more aggressive block production reduction to
+save the cluster. For example, a leader could produce only one block in four
+consecutive slots allocated to it.
+
+However, this only solves the problem in specific type of outage, and it seems
+risky to aggressively reduce block production, so we are not proceeding with
+this proposal for now.
+
+## Detailed Design
+
+The new protocol tries to make all restarting validators get the same data
+blocks and the same set of last votes, then they will almost certainly make the 
```

Comment:
`then they will almost certainly` -> `so that they will with high probability converge on the same`

Reply:
Changed